File: scala/src/main/scala/com/gildedrose/BackstageItem.scala
package com.gildedrose

class BackstageItem(item: Item) extends BaseItem(item) {

  override protected def updateQuality: Unit = {
    increaseQuality
    if (item.sellIn < 10) {
      increaseQuality
    }
    if (item.sellIn < 5) {
      increaseQuality
    }
    if (item.sellIn < 0) {
      item.quality = 0
    }
  }

  private def increaseQuality = {
    if (item.quality < 50) {
      item.quality += 1
    }
  }
}

object BackstageItem extends BaseFactory {

  def create(item: Item): Option[BaseItem] = if ("Backstage passes to a TAFKAL80ETC concert".equals(item.name)) Option(new BackstageItem(item)) else None;

}

File: Java/src/main/java/com/gildedrose/BackstageItem.java
package com.gildedrose;


public class BackstageItem extends BaseItem {

    protected BackstageItem(String name, int sellIn, int quality) {
        super(name, sellIn, quality);
    }

    @Override
    public void updateQuality() {
        if (this.quality < 50) {
            this.quality++;
        }
        if (this.sellIn < 10) {
            if (this.quality < 50) {
                this.quality++;
            }
        }

        if (this.sellIn < 5) {
            if (this.quality < 50) {
                this.quality++;
            }
        }
        if (this.sellIn < 0) {
            this.quality = 0;
        }
    }

}
